pkg/controller/datavolume-controller.go

```diff
@@ -2599,3 +2623,288 @@ func GetRequiredSpace(filesystemOverhead float64, requestedSpace int64) int64 {
 func newLongTermCloneTokenGenerator(key *rsa.PrivateKey) token.Generator {
 	return token.NewGenerator(common.ExtendedCloneTokenIssuer, key, 10*365*24*time.Hour)
 }
+
+// detectCloneSize obtains and assigns the original PVC's size when cloning using an empty storage value
+func (r *DatavolumeReconciler) detectCloneSize(
+	dv *cdiv1.DataVolume,
+	pvcSpec *corev1.PersistentVolumeClaimSpec,
+	cloneType cloneStrategy) (bool, error) {
+
+	var targetSize resource.Quantity
+	sourcePvc, err := r.findSourcePvc(dv)
+	if err != nil {
+		return false, err
+	}
+	sourceSize := sourcePvc.Status.Capacity.Storage()
+
+	// Due to possible filesystem overhead complications when cloning
+	// using host-assisted strategy, we create a pod that automatically
+	// collects the size of the original virtual image with 'qemu-img'.
+	// If another strategy is used or the original PVC's volume mode
+	// is "block", we simply extract the value from the original PVC's spec.
+	if cloneType == HostAssistedClone && getVolumeMode(sourcePvc) == corev1.PersistentVolumeFilesystem {
+		// If available, we first try to get the virtual size from previous iterations
+		imgSize, available := getSizeFromAnnotations(sourcePvc)
+		if !available {
+			imgSize, err = r.getSizeFromPod(sourcePvc)
+			if err != nil {
+				return false, err
+			} else if imgSize == ImgSizeNotReady {
+				return false, nil
+			}
+		}
+		// We enable the clone-controller to skip the size comparison requirement
+		dv.Annotations[AnnPermissiveClone] = "true"
+		// If needed, inflate size with filesystem overhead
+		targetSize, err = inflateSizeWithOverhead(r.client, imgSize, pvcSpec, sourceSize)
+		if err != nil {
+			return false, err
+		}
+	} else {
+		targetSize = *sourceSize
+	}
+
+	pvcSpec.Resources.Requests[corev1.ResourceStorage] = targetSize
+	return true, nil
+}
+
+// getSizeFromAnnotations checks the source PVC's annotations and returns the requested size if it has already been obtained
+func getSizeFromAnnotations(sourcePvc *corev1.PersistentVolumeClaim) (string, bool) {
+	virtualImageSize, available := sourcePvc.Annotations[AnnVirtualImageSize]
+	if available == true {
+		sourceCapacity, available := sourcePvc.Annotations[AnnSourceCapacity]
+		currCapacity := sourcePvc.Status.Capacity
+		// Checks if the original PVC's capacity has changed
+		if available && currCapacity.Storage().Cmp(resource.MustParse(sourceCapacity)) == 0 {
+			return virtualImageSize, true
+		}
+	}
+
+	return "0", false
+}
+
+// getSizeFromPod attempts to get the image size from a pod that directly obtains said value from the source PVC
+func (r *DatavolumeReconciler) getSizeFromPod(sourcePvc *corev1.PersistentVolumeClaim) (string, error) {
+	// The pod should not be created until the source PVC has finished the import process
+	if !hasPvcSucceeded(sourcePvc) {
+		return ImgSizeNotReady, nil
```

Comment:
I like that you are differentiating between pod is not ready type errors and other errors. But returning that as part of the actual value is probably not what we want, especially if we decide to return a number. It might make sense to return your own error type so in a higher up function you can decide to do something special based on this special error type. See https://github.com/kubevirt/containerized-data-importer/blob/main/pkg/importer/data-processor.go#L65-L70 for an example of how to make a custom error.

Reply:
Yeah, that makes more sense. Thanks!